File: client/python/ff_client/ff_client.py
from .ff_config import FfConfig
from .ff_request import FfRequest
from .udp_packet import UdpPacket
import socket
import struct
import logging
import time
from Cryptodome.Random import get_random_bytes, random
from Cryptodome.Cipher import AES
from Cryptodome.Protocol.KDF import PBKDF2
from Cryptodome.Hash import SHA256
from typing import List


class FfClient:
    MAX_PACKET_LENGTH = 1300
    OPTION_HEADER_LENGTH = 3

    def __init__(self, config: FfConfig):
        self.config = config
        self.logger = self.init_default_logger()

    def init_default_logger(self):
        logger = logging.getLogger('ff')
        logger.setLevel(self.config.log_level)
        ch = logging.StreamHandler()
        formatter = logging.Formatter(
            '%(asctime)s - %(name)s - %(levelname)s - %(message)s')
        ch.setFormatter(formatter)

        logger.addHandler(ch)

        return logger

    def send_request(self, http_request: str, https: bool = True):
        packets = self.create_request_packets(http_request, https)

        self.logger.info('Creating socket')

        with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
            self.logger.info('Sending %d packets to %s:%s' % (
                len(packets), self.config.ip_address, self.config.port))

            for packet in packets:
                sock.sendto(packet.payload[:packet.length],
                            (self.config.ip_address, self.config.port))
            
            self.logger.info('Packets sent!')

    def create_request_packets(self, http_request: str, https: bool = True):
        self.logger.debug('Initialising ff request')
        request = FfRequest(version=FfRequest.Version.V1,
                            request_id=random.getrandbits(64))

        self.create_secure_options(request, https)
        secure_options = self.serialize_secure_options(request.secure_options)
        payload = secure_options + bytearray(http_request.encode('utf8'))

        if (self.config.pre_shared_key):
            payload = self.encrypted_request_payload(
                request, payload)

        request.options.append(FfRequest.Option(
            FfRequest.Option.Type.BREAK, 0, bytearray([])))

        request.payload = payload

        return self.packetise_request(request)

    def create_secure_options(self, request: FfRequest, https: bool):
        if https:
            request.secure_options.append(FfRequest.Option(
                FfRequest.Option.Type.HTTPS, 1, bytearray([1])))

        request.secure_options.append(FfRequest.Option(
            FfRequest.Option.Type.TIMESTAMP, 8, self.get_current_timestamp_bytes()))

        request.secure_options.append(FfRequest.Option(
            FfRequest.Option.Type.EOL, 0, bytearray([])))

    def get_current_timestamp_bytes(self):
        timestamp = int(time.time())
        buff = bytearray(8)

        self.pack_into_size('!Q', buff, 0, timestamp)

        return buff

    def serialize_secure_options(self, options: List[FfRequest.Option]) -> bytearray:
        length = sum(map(lambda i: self.OPTION_HEADER_LENGTH + i.length, options))
        buff = bytearray(length)

        self.write_request_options(buff, 0, options)
        print(buff)
        return buff

    def encrypted_request_payload(self, request: FfRequest, payload: bytearray) -> str:
        self.logger.debug('Encrypting request payload')

        if not self.config.pre_shared_key:
            raise RuntimeError(
                'Cannot encrypt payload without pre_shared_key set')

        salt = get_random_bytes(16)
        derived_key = PBKDF2(self.config.pre_shared_key, salt, 
                             dkLen=32,
                             count=self.config.pbkdf2_iterations, 
                             hmac_hash_module=SHA256)

        iv = get_random_bytes(12)

        cipher = AES.new(key=derived_key, mode=AES.MODE_GCM,
                         nonce=iv, mac_len=16)

        ciphertext, tag = cipher.encrypt_and_digest(payload)

        request.options.append(FfRequest.Option(
            FfRequest.Option.Type.ENCRYPTION_MODE,
            1,
            bytearray([FfRequest.EncryptionMode.AES_256_GCM])
        ))

        request.options.append(FfRequest.Option(
            FfRequest.Option.Type.ENCRYPTION_IV,
            len(iv),
            bytearray(iv)
        ))

        request.options.append(FfRequest.Option(
            FfRequest.Option.Type.ENCRYPTION_TAG,
            len(tag),
            bytearray(tag)
        ))

        request.options.append(FfRequest.Option(
            FfRequest.Option.Type.KEY_DERIVE_MODE,
            1,
            bytearray([FfRequest.KeyDeriveMode.PBKDF2])
        ))

        request.options.append(FfRequest.Option(
            FfRequest.Option.Type.KEY_DERIVE_SALT,
            len(salt),
            bytearray(salt)
        ))

        self.logger.debug('Encrypted request into %d bytes' % len(ciphertext))
        return ciphertext

    def packetise_request(self, request: FfRequest) -> List[UdpPacket]:
        self.logger.debug('Packetising request')

        if isinstance(request.payload, bytearray):
            payload = request.payload
        elif isinstance(request.payload, str):
            payload = bytearray(request.payload.encode('utf8'))
        elif isinstance(request.payload, bytes):
            payload = bytearray(request.payload)
        else:
            raise ValueError(
                'Unknown request payload type, expecting bytearray, str or bytes, got %s' % request.payload)

        packets = []
        chunk_offset = 0
        bytes_left = len(payload)

        while(bytes_left > 0):
            packet_buff = bytearray(self.MAX_PACKET_LENGTH)
            ptr = 0

            # @see https://docs.python.org/2/library/struct.html#format-characters
            ptr += self.pack_into_size('!H', packet_buff, ptr, request.version)
            ptr += self.pack_into_size('!Q', packet_buff,
                                       ptr, request.request_id)
            ptr += self.pack_into_size('!I', packet_buff,
                                       ptr, len(request.payload))
            ptr += self.pack_into_size('!I', packet_buff, ptr, chunk_offset)
            # defer writing chunk length
            chunk_length_ptr = ptr
            ptr += 2

            options = request.options if chunk_offset == 0 else [
                FfRequest.Option(FfRequest.Option.Type.EOL, 0, bytearray([]))]

            ptr = self.write_request_options(packet_buff, ptr, options)

            chunk_length = min(self.MAX_PACKET_LENGTH - ptr, bytes_left)
            self.pack_into_size('!H', packet_buff,
                                chunk_length_ptr, chunk_length)

            for i in range(chunk_offset, chunk_offset + chunk_length):
                packet_buff[ptr] = payload[i]
                ptr += 1

            bytes_left -= chunk_length
            chunk_offset += chunk_length

            packets.append(UdpPacket(ptr, packet_buff))

        self.logger.debug('Packetised request into %d packets' % len(packets))

        return packets

    def write_request_options(self, buff: bytearray, ptr: int, options: List[FfRequest.Option]) -> int:
        for option in options:
            ptr += self.pack_into_size('!B', buff, ptr, option.type)
            ptr += self.pack_into_size('!H',
                                        buff, ptr, option.length)
            for i in option.value:
                buff[ptr] = i
                ptr += 1

        return ptr

    def pack_into_size(self, format: str, buff: bytearray, offset: int, value) -> int:
        size = struct.calcsize(format)
        struct.pack_into(format, buff, offset, value)

        return size

File: client/java/src/main/java/com/timetoogo/ffclient/FfClient.java

package com.timetoogo.ffclient;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse.BodySubscriber;
import java.net.http.HttpResponse.BodySubscribers;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.Map.Entry;
import java.util.concurrent.Flow;
import java.util.logging.Logger;

import javax.crypto.Cipher;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;

import java.nio.ByteBuffer;
import java.nio.charset.Charset;
import java.security.SecureRandom;

public class FfClient {
    private final Logger logger = Logger.getLogger(FfClient.class.getName());
    private final Charset UTF8 = Charset.forName("UTF-8");
    private FfConfig config;
    private final int MAX_PACKET_LENGTH = 1300;

    static final class ByteArraySubscriber implements Flow.Subscriber<ByteBuffer> {
        final BodySubscriber<byte[]> wrapped;

        private ByteArraySubscriber(BodySubscriber<byte[]> wrapped) {
            this.wrapped = wrapped;
        }

        @Override
        public void onSubscribe(Flow.Subscription subscription) {
            wrapped.onSubscribe(subscription);
        }

        @Override
        public void onNext(ByteBuffer item) {
            wrapped.onNext(List.of(item));
        }

        @Override
        public void onError(Throwable throwable) {
            wrapped.onError(throwable);
        }

        @Override
        public void onComplete() {
            wrapped.onComplete();
        }
    }

    public FfClient(FfConfig config) {
        this.config = config;
    }

    public void sendRequest(HttpRequest httpRequest) throws Exception {
        var packets = this.createRequestPackets(httpRequest);

        this.logger.info("Creating socket");
        var socket = new DatagramSocket(0, InetAddress.getLocalHost());

        socket.setReuseAddress(true);
        socket.connect(this.config.getIpAddress(), this.config.getPort());

        for (var packet : packets) {
            socket.send(new DatagramPacket(packet.getValue(), packet.getLength()));
        }

        this.logger.info(String.format("Sent %d packets to %s:%d", packets.size(),
                this.config.getIpAddress().toString(), this.config.getPort()));

        socket.close();
    }

    List<UdpPacket> createRequestPackets(HttpRequest httpRequest) throws Exception {
        FfRequest request = new FfRequest();
        request.setVersion(FfRequest.Version.V1);
        request.setRequestId(this.getRandomInstance().nextLong());

        if (httpRequest.uri().getScheme().equalsIgnoreCase("https")) {
            request.getOptions().add(FfRequestOption.builder().type(FfRequestOption.Type.HTTPS)
                    .length((short) 1).value(new byte[] {1}).build());
        }

        byte[] httpMessage = this.serializeHttpRequest(httpRequest);

        if (this.config.getPreSharedKey() != null && !this.config.getPreSharedKey().isEmpty()) {
            httpMessage = this.encryptHttpMessage(httpMessage, request);
        }

        request.setPayload(httpMessage);

        request.getOptions().add(
                FfRequestOption.builder().type(FfRequestOption.Type.EOL).length((short) 0).build());

        return this.packetiseRequest(request);
    }

    byte[] serializeHttpRequest(HttpRequest request) throws IOException {
        this.logger.info("Serializing HTTP request");
        ByteArrayOutputStream buffer = new ByteArrayOutputStream();

        buffer.write(request.method().getBytes(UTF8));
        buffer.write(" ".getBytes(UTF8));
        buffer.write(
                (request.uri().getPath().isEmpty() ? "/" : request.uri().getPath()).getBytes(UTF8));

        if (request.uri().getQuery() != null && !request.uri().getQuery().isEmpty()) {
            buffer.write("?".getBytes(UTF8));
            buffer.write(request.uri().getQuery().getBytes(UTF8));
        }

        buffer.write(" HTTP/1.1\n".getBytes(UTF8));

        if (request.uri().getHost() != null) {
            buffer.write("Host: ".getBytes(UTF8));
            buffer.write(request.uri().getHost().getBytes(UTF8));
            buffer.write("\n".getBytes(UTF8));
        }

        if (request.bodyPublisher().isPresent()) {
            var contentLength = request.bodyPublisher().get().contentLength();
            buffer.write("Content-Length: ".getBytes(UTF8));
            buffer.write(Long.toString(contentLength).getBytes(UTF8));
            buffer.write("\n".getBytes(UTF8));
        }

        for (Entry<String, List<String>> header : request.headers().map().entrySet()) {
            for (String value : header.getValue()) {
                buffer.write(header.getKey().getBytes(UTF8));
                buffer.write(": ".getBytes(UTF8));
                buffer.write(value.getBytes(UTF8));
                buffer.write("\n".getBytes(UTF8));
            }
        }

        buffer.write("\n".getBytes(UTF8));

        if (request.bodyPublisher().isPresent()) {
            var bodySubscriber = BodySubscribers.ofByteArray();
            var byteArraySubscriber = new ByteArraySubscriber(bodySubscriber);
            request.bodyPublisher().get().subscribe(byteArraySubscriber);
            byte[] body = bodySubscriber.getBody().toCompletableFuture().join();

            buffer.write(body);
        }

        byte[] bytes = buffer.toByteArray();

        this.logger.info(String.format("Serialized request to %d bytes", bytes.length));

        return bytes;
    }

    byte[] encryptHttpMessage(byte[] httpMessage, FfRequest request) throws Exception {
        this.logger.info("Encrypting payload");
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");

        var salt = new byte[16];
        this.getRandomInstance().nextBytes(salt);

        var spec = new PBEKeySpec(this.config.getPreSharedKey().toCharArray(), salt,
                this.config.getPbkdf2Iterations(), 256);
        var skf = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
        var derivedKey = skf.generateSecret(spec).getEncoded();
        var keySpec = new SecretKeySpec(derivedKey, "AES");

        var iv = new byte[12];
        this.getRandomInstance().nextBytes(iv);

        var tagLength = 16;
        var gcmSpec = new GCMParameterSpec(tagLength * 8, iv);

        cipher.init(Cipher.ENCRYPT_MODE, keySpec, gcmSpec);

        byte[] cipherTextWithTag = cipher.doFinal(httpMessage);

        byte[] cipherText = new byte[cipherTextWithTag.length - tagLength];
        byte[] tag = new byte[tagLength];
        System.arraycopy(cipherTextWithTag, 0, cipherText, 0, cipherText.length);
        System.arraycopy(cipherTextWithTag, cipherText.length, tag, 0, tag.length);

        this.logger.info(String.format("Encrypted message into %d bytes", cipherText.length));

        request.getOptions().add(FfRequestOption.builder()
                .type(FfRequestOption.Type.ENCRYPTION_MODE).length((short) 1)
                .value(new byte[] {FfRequest.EncryptionMode.AES_256_GCM.getValue()}).build());

        request.getOptions().add(FfRequestOption.builder().type(FfRequestOption.Type.ENCRYPTION_IV)
                .length((short) iv.length).value(iv).build());

        request.getOptions().add(FfRequestOption.builder().type(FfRequestOption.Type.ENCRYPTION_TAG)
                .length((short) tag.length).value(tag).build());

        request.getOptions()
                .add(FfRequestOption.builder().type(FfRequestOption.Type.KEY_DERIVE_MODE)
                        .length((short) 1)
                        .value(new byte[] {FfRequest.KeyDeriveMode.PBKDF2.getValue()}).build());

        request.getOptions()
                .add(FfRequestOption.builder().type(FfRequestOption.Type.KEY_DERIVE_SALT)
                        .length((short) salt.length).value(salt).build());

        return cipherText;
    }

    List<UdpPacket> packetiseRequest(FfRequest request) {
        this.logger.info("Packetising request");
        var packets = new ArrayList<UdpPacket>();

        int chunkOffset = 0;
        int bytesLeft = request.getPayload().length;

        while (bytesLeft > 0) {
            byte[] packetBuff = new byte[MAX_PACKET_LENGTH];
            int ptr = 0;

            ptr = this.writeShort(packetBuff, ptr, request.getVersion().getValue());
            ptr = this.writeLong(packetBuff, ptr, request.getRequestId());
            ptr = this.writeInt(packetBuff, ptr, request.getPayload().length);
            ptr = this.writeInt(packetBuff, ptr, chunkOffset);
            // Defer writing chuck length
            int chunkLengthPtr = ptr;
            ptr += 2;

            var options =
                    chunkOffset == 0 ? request.getOptions() : new ArrayList<FfRequestOption>() {
                        {
                            add(FfRequestOption.builder().type(FfRequestOption.Type.EOL)
                                    .length((short) 0).build());
                        }
                    };

            for (var option : options) {
                packetBuff[ptr++] = option.getType().getValue();
                ptr = this.writeShort(packetBuff, ptr, option.getLength());

                if (option.getLength() > 0) {
                    System.arraycopy(option.getValue(), 0, packetBuff, ptr, option.getLength());
                    ptr += option.getLength();
                }
            }

            short chunkLength = (short) Math.min(MAX_PACKET_LENGTH - ptr, bytesLeft);

            this.writeShort(packetBuff, chunkLengthPtr, chunkLength);

            System.arraycopy(request.getPayload(), chunkOffset, packetBuff, ptr, chunkLength);
            ptr += chunkLength;

            chunkOffset += chunkLength;
            bytesLeft -= chunkLength;

            packets.add(new UdpPacket(packetBuff, ptr));
        }

        this.logger.info(String.format("Converted request into %d packets", packets.size()));

        return packets;
    }

    private int writeShort(byte[] buff, int offset, short val) {
        int intVal = Short.toUnsignedInt(val);
        buff[offset++] = (byte) ((intVal & 0xff00) >>> 8);
        buff[offset++] = (byte) (intVal & 0xff);

        return offset;
    }

    private int writeInt(byte[] buff, int offset, int val) {
        long longVal = Integer.toUnsignedLong(val);
        buff[offset++] = (byte) ((longVal & 0xff000000) >>> 24);
        buff[offset++] = (byte) ((longVal & 0xff0000) >>> 16);
        buff[offset++] = (byte) ((longVal & 0xff00) >>> 8);
        buff[offset++] = (byte) (longVal & 0xff);

        return offset;
    }

    private int writeLong(byte[] buff, int offset, long val) {
        buff[offset++] = (byte) ((val & 0xff00000000000000L) >>> 56);
        buff[offset++] = (byte) ((val & 0xff000000000000L) >>> 48);
        buff[offset++] = (byte) ((val & 0xff0000000000L) >>> 40);
        buff[offset++] = (byte) ((val & 0xff00000000L) >>> 36);
        buff[offset++] = (byte) ((val & 0xff000000) >>> 24);
        buff[offset++] = (byte) ((val & 0xff0000) >>> 16);
        buff[offset++] = (byte) ((val & 0xff00) >>> 8);
        buff[offset++] = (byte) (val & 0xff);

        return offset;
    }

    private Random getRandomInstance() throws Exception {
        if (System.getenv("FF_WEAK_RANDOM") != null) {
            this.logger.warning("Using weak RNG");
            return new Random(System.currentTimeMillis());
        } else {
            return SecureRandom.getInstanceStrong();
        }
    }
}
